### ISSUE TITLE
irmap cache doesn't work for overlayfs

### ISSUE THREAD

**Issue by avagin** [checkpoint-restore/criu#136]
```
===================== Run zdtm/static/inotify_irmap in ns ======================
Start test
Test is SUID
./inotify_irmap --pidfile=inotify_irmap.pid --outfile=inotify_irmap.out
Run criu dump
=[log]=> dump/zdtm/static/inotify_irmap/53/1/dump.log
------------------------ grep Error ------------------------
(00.007855) Error (fsnotify.c:215): fsnotify: Can't open file handle for 0x00000029:0x0000000000143439: Operation not permitted
(00.007897) Error (irmap.c:86): irmap: Can't stat /var/spool: No such file or directory
(00.007903) Error (irmap.c:86): irmap: Can't stat /var/log: No such file or directory
(00.007920) Error (irmap.c:86): irmap: Can't stat /usr/share/dbus-1/system-services: No such file or directory
(00.007925) Error (irmap.c:86): irmap: Can't stat /var/lib/polkit-1/localauthority: No such file or directory
(00.007931) Error (irmap.c:86): irmap: Can't stat /usr/share/polkit-1/actions: No such file or directory
(00.007937) Error (irmap.c:86): irmap: Can't stat /lib/udev: No such file or directory
(00.007946) Error (irmap.c:86): irmap: Can't stat /no-such-path: No such file or directory
(00.007948) Error (fsnotify.c:285): fsnotify:   Can't dump that handle
(00.007959) Error (cr-dump.c:1303): Dump files (pid: 56) failed with -1
(00.010536) Error (cr-dump.c:1605): Dumping FAILED.
------------------------ ERROR OVER ------------------------
############### Test zdtm/static/inotify_irmap FAIL at CRIU dump ###############
Wait for zdtm/static/inotify_irmap to die for 0.100000
##################################### FAIL #####################################
```

```
28997 pts/0    S+     0:00  |               \_ python test/zdtm.py run -t zdtm/static/inotify_irmap --sbs -f ns
29007 pts/0    S+     0:00  |                   \_ ./zdtm_ct zdtm.py
29009 pts/0    S+     0:00  |                       \_ python2 zdtm.py
29011 pts/0    S+     0:00  |                           \_ python2 zdtm.py
29038 ?        Ss     0:00  |                           \_ ./inotify_irmap --pidfile=inotify_irmap.pid --outfile=inotify_irmap.out
29042 ?        Ss     0:00  |                               \_ ./inotify_irmap --pidfile=inotify_irmap.pid --outfile=inotify_irmap.out
```

```
29042 inotify_add_watch(5, "/etc/zdtm-test", IN_OPEN) = 1
29042 inotify_add_watch(5, "/etc/zdtm-test1", IN_OPEN) = 2
```

```
[root@fc22-vm criu]# cat /proc/29042/fdinfo/5 
pos:    0
flags:  04000
mnt_id: 11
inotify wd:2 ino:142c7d sdev:29 mask:800a020 ignored_mask:0 fhandle-bytes:8 fhandle-type:1 f_handle:7d2c140000000000
inotify wd:1 ino:142c7c sdev:29 mask:800a020 ignored_mask:0 fhandle-bytes:8 fhandle-type:1 f_handle:7c2c140000000000

[root@fc22-vm criu]# stat test/etc/zdtm-test
  File: ‘test/etc/zdtm-test’
  Size: 0           Blocks: 0          IO Block: 4096   regular empty file
Device: fc02h/64514d    Inode: 18213       Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2016-03-11 19:33:14.101000000 +0300
Modify: 2016-03-11 19:33:14.101000000 +0300
Change: 2016-03-11 19:33:14.101000000 +0300
 Birth: -
[root@fc22-vm criu]# stat test.up/etc/zdtm-test
  File: ‘test.up/etc/zdtm-test’
  Size: 0           Blocks: 0          IO Block: 4096   regular empty file
Device: fc02h/64514d    Inode: 18213       Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2016-03-11 19:33:14.101000000 +0300
Modify: 2016-03-11 19:33:14.101000000 +0300
Change: 2016-03-11 19:33:14.101000000 +0300
 Birth: -
[root@fc22-vm criu]# stat test/
  File: ‘test/’
  Size: 4096        Blocks: 8          IO Block: 4096   directory
Device: 29h/41d Inode: 15116       Links: 1
Access: (0777/drwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2016-03-11 19:20:21.625000000 +0300
Modify: 2016-03-11 18:59:26.996000000 +0300
Change: 2016-03-11 18:59:26.996000000 +0300
 Birth: -
```

**Comment by avagin** [comment]
[root@fc22-vm criu]# ls -l /proc/31098/fd
total 0
lrwx------ 1 root root 64 Mar 11 20:08 0 -> /dev/null
l-wx------ 1 root root 64 Mar 11 20:08 1 -> /zdtm/static/inotify_irmap.out.inprogress
l-wx------ 1 root root 64 Mar 11 20:08 2 -> /zdtm/static/inotify_irmap.out.inprogress
l-wx------ 1 root root 64 Mar 11 20:08 3 -> /etc/zdtm-test
l-wx------ 1 root root 64 Mar 11 20:08 4 -> /etc/zdtm-test1
lr-x------ 1 root root 64 Mar 11 20:08 5 -> anon_inode:inotify
[root@fc22-vm criu]# 
[root@fc22-vm criu]# cat /proc/31098/fdinfo/3 
pos:	0
flags:	0100001
mnt_id:	156
[root@fc22-vm criu]# cat /proc/31098/mountinfo 
156 120 0:41 / / rw,relatime - overlay overlay rw,lowerdir=test,upperdir=test.up,workdir=test.work
157 156 0:47 / /proc rw,nosuid,nodev,noexec,relatime - proc proc rw
121 156 0:48 / /dev/pts rw,relatime - devpts pts rw,mode=666,ptmxmode=666
[root@fc22-vm criu]# stat -L /proc/31098/fd/3
  File: ‘/proc/31098/fd/3’
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: fc02h/64514d	Inode: 18213       Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2016-03-11 19:44:21.422000000 +0300
Modify: 2016-03-11 19:44:21.422000000 +0300
Change: 2016-03-11 19:44:21.422000000 +0300
 Birth: -

**Comment by lynkz-matt-psaltis** [comment]
Anyone have a workaround for this? Trying to make this work on wsl2 :D

Is a reopen possible?

**Comment by rppt** [comment]
I'm not sure anyone is working on this, but reopen is possible

**Comment by mihalicyn** [comment]
Hi @lynkz-matt-psaltis 

Since linux kernel version ~4.16:
- https://github.com/torvalds/linux/commit/8383f1748829e1a6a07988863ed13a47fb653387
- https://github.com/torvalds/linux/commit/8ed5eec9d6c4c013aa657ebefbd10a1a0d15893d

overlayfs supports file handles. So, you need to mount your overlayfs with `nfs_export=on` parameter to make this work. You can also make this behavior default for all overlayfs mounts by `echo Y | sudo tee /sys/module/overlay/parameters/nfs_export`.

irmap relies on file handles, so it should help.

**Comment by 2000yeshu** [comment]
Did anyone test this solution. Does this work?

**Comment by avagin** [comment]
This issue was fixed in the 4.2 kernel;
commit 4bacc9c9234c7c8eec44f5ed4e960d9f96fa0f01
Author: David Howells <dhowells@redhat.com>
Date:   Thu Jun 18 14:32:31 2015 +0100

    overlayfs: Make f_path always point to the overlay and f_inode to the underlay
